onnxruntime/core/framework/sparse_utils.cc

```diff
@@ -0,0 +1,510 @@
+  }
+
+  const bool is_string = src.IsDataTypeString();
+
+  if (is_string && dst_allocator->Info().device.Type() != OrtDevice::CPU) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                           "Unable to convert strings tensor to a sparse tensor that is not on CPU");
+  }
+
+  const AllocatorPtr& conversion_allocator = (dst_allocator->Info().device.Type() == OrtDevice::CPU)
+                                                 ? dst_allocator
+                                                 : cpu_allocator;
+
+  Tensor cpu_result(src.DataType(), src.Shape(), conversion_allocator);
+  if (!is_string) {
+    memset(cpu_result.MutableDataRaw(), 0, cpu_result.SizeInBytes());
+  }
+
+  if (src.NumValues() > 0) {
+    const auto rows = src_dims[0];
+    const auto cols = src_dims[1];
+
+    {
+      auto csr_view = src.AsCsr();
+      const auto inner_num = csr_view.Inner().Shape().Size();
+      const auto outer_num = csr_view.Outer().Shape().Size();
+      ORT_ENFORCE(inner_num == src.Values().Shape().Size(), "Expecting inner indices to be same as nnz. Got: ", inner_num);
+      ORT_ENFORCE(outer_num == (rows + 1), "Outer indices must be M + 1. Got: ", outer_num);
+    }
+
+    CopyElementFunc copy_func;
+    if (is_string) {
+      copy_func = CopyString;
+    } else {
+      const auto element_size = src.DataType()->Size();
+      switch (element_size) {
+        case sizeof(uint8_t):
+          copy_func = CopyElement<uint8_t>;
+          break;
+        case sizeof(uint16_t):
+          copy_func = CopyElement<uint16_t>;
+          break;
+        case sizeof(uint32_t):
+          copy_func = CopyElement<uint32_t>;
+          break;
+        case sizeof(uint64_t):
+          copy_func = CopyElement<uint64_t>;
+          break;
+        default:
+          return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Unsupported element size: ", element_size);
+      }
+    }
+
+    SparseTensor cpu_src;
+    const void* values = nullptr;
+    gsl::span<const int64_t> inner_span;
+    gsl::span<const int64_t> outer_span;
+    if (src.Location().device.Type() != OrtDevice::CPU) {
+      SparseTensor t(src.DataType(), src.Shape(), cpu_allocator);
+      ORT_RETURN_IF_ERROR(data_manager.CopyTensor(src, t));
+      cpu_src = std::move(t);
+      values = cpu_src.Values().DataRaw();
+      inner_span = cpu_src.AsCsr().Inner().DataAsSpan<int64_t>();
+      outer_span = cpu_src.AsCsr().Outer().DataAsSpan<int64_t>();
+    } else {
+      values = src.Values().DataRaw();
+      inner_span = src.AsCsr().Inner().DataAsSpan<int64_t>();
+      outer_span = src.AsCsr().Outer().DataAsSpan<int64_t>();
+    }
+
+    void* output = cpu_result.MutableDataRaw();
+
+    size_t src_idx = 0;
+    size_t inner_idx = 0;
+    for (size_t out_i = 1; out_i < outer_span.size(); ++out_i) {
+      auto row_size = outer_span[out_i] - outer_span[out_i - 1];
+      for (int64_t cnt = 0; cnt < row_size; ++cnt, ++inner_idx) {
```

Comment:
Do we need separate tests to keep these utils tested ?

Reply:
Yes, in this PR